crates/core/src/geoarrow/mod.rs

```diff
@@ -250,12 +274,107 @@ pub fn add_wkb_metadata(mut record_batch: RecordBatch, column_name: &str) -> Res
+        Point(typ) => {
+            let builder = PointBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        LineString(typ) => {
+            let builder = LineStringBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        Polygon(typ) => {
+            let builder = PolygonBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        MultiPoint(typ) => {
+            let builder = MultiPointBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        MultiLineString(typ) => {
+            let builder = MultiLineStringBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        MultiPolygon(typ) => {
+            let builder = MultiPolygonBuilder::from_nullable_geometries(&geoms, typ)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        GeometryCollection(typ) => {
+            let builder =
+                GeometryCollectionBuilder::from_nullable_geometries(&geoms, typ, prefer_multi)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        Rect(_) => {
+            Err(GeoArrowError::General(format!("Unexpected data type {:?}", target_type,)).into())
+        }
+        Geometry(typ) => {
+            let builder = GeometryBuilder::from_nullable_geometries(&geoms, typ, prefer_multi)?;
+            Ok(Arc::new(builder.finish()))
+        }
+        _ => todo!("Handle target WKB/WKT in `from_wkb`"),
+    }
+}
+
+/// Convert to WKB
+fn to_wkb<O: OffsetSizeTrait>(arr: &dyn GeoArrowArray) -> Result<WkbArray<O>> {
+    use GeoArrowType::*;
+    match arr.data_type() {
+        Point(_) => impl_to_wkb(arr.as_point()),
+        LineString(_) => impl_to_wkb(arr.as_line_string()),
+        Polygon(_) => impl_to_wkb(arr.as_polygon()),
+        MultiPoint(_) => impl_to_wkb(arr.as_multi_point()),
+        MultiLineString(_) => impl_to_wkb(arr.as_multi_line_string()),
+        MultiPolygon(_) => impl_to_wkb(arr.as_multi_polygon()),
+        Geometry(_) => impl_to_wkb(arr.as_geometry()),
+        GeometryCollection(_) => impl_to_wkb(arr.as_geometry_collection()),
+        Rect(_) => impl_to_wkb(arr.as_rect()),
+        Wkb(_) => impl_to_wkb(arr.as_wkb::<i32>()),
+        LargeWkb(_) => impl_to_wkb(arr.as_wkb::<i64>()),
+        Wkt(_) => impl_to_wkb(arr.as_wkt::<i32>()),
+        LargeWkt(_) => impl_to_wkb(arr.as_wkt::<i64>()),
+    }
+}
+
+fn impl_to_wkb<'a, O: OffsetSizeTrait>(geo_arr: &'a impl ArrayAccessor<'a>) -> Result<WkbArray<O>> {
+    let metadata = geo_arr.data_type().metadata().clone();
+
+    let geoms = geo_arr
+        .iter()
+        .map(|x| x.transpose())
+        .collect::<std::result::Result<Vec<_>, _>>()?;
+    let wkb_type = WkbType::new(metadata);
+    Ok(WkbBuilder::from_nullable_geometries(geoms.as_slice(), wkb_type).finish())
+}
+
```

Comment:
This all can be removed post https://github.com/geoarrow/geoarrow-rs/pull/1070. 

In case you have an idea on the API (https://github.com/geoarrow/geoarrow-rs/pull/1070#discussion_r2050996867) lmk 😄